### ISSUE TITLE
Extension mechanism - merging in

### ISSUE THREAD

**Issue by MarcinZukowski** [atari800/atari800#134]
Hi,

For a while I've been working on an "extension" mechanism for atari800. It has been discussed [here](https://atariage.com/forums/topic/321523-some-fun-with-atari800/) (English) and [here](http://www.atari.org.pl/forum/viewtopic.php?pid=281264#p281264) (Polish).

There's a short [README](https://github.com/MarcinZukowski/atari800/blob/better-yoomp/data/ext/README.md) available. 

You can see the code changes here: https://github.com/atari800/atari800/compare/master...MarcinZukowski:better-yoomp?expand=1

I would like to merge this back into the main repo. However, it's quite a lot of code, and quite a few changes. So I wanted to first discuss it with the atari800 maintainers.

Few things I wanted to raise:
* the code uses some C99 stuff - are we OK moving atari800 to use it? It's 2021 after all. Here's the list I'm using to compile: 
  ```
    -Wno-c99-extensions \
    -Wno-declaration-after-statement -Wno-missing-prototypes \
    -Wno-gcc-compat -Wno-long-long -Wno-comment -Wno-variadic-macros \
    -Werror=incompatible-function-pointer-types"
  ```
  I could probably make the code work with the options as currently, but it wouldn't be fun :)
* there are a few main things in the code, that could be moved into atari800 in stages
  * changes to the existing atari800 code (e.g. adding `--with-sound=sdlmixer` or extracting `src/sdl/video_gl-common.h`)
  * extension hooks in the existing code (e.g. in [cpu.c](https://github.com/atari800/atari800/compare/master...MarcinZukowski:better-yoomp?expand=1#diff-ef499f81601b94a0e616ca923ac08b37fe84440f3dd67e25b52713b83d9b9cc6))
  * the extension framework itself (mostly `src/ext.*` and `src/sdl/video_gl-ext.*`)
  * actual extensions in C (`src/ext/*`)
  * actual extensions in Lua (`data/ext/*`)

If you think it's worth bringing it into the main repo, we could do it all in one go (extensions are disabled by default), or in stages.

In short, just wanted to start the discussion, as I'm not sure how best to proceed.

Thanks!

**Comment by joysfera** [comment]
Anyone has an idea which platforms get unsupported if we move to C99?

If we agree to do that then I'd like to do one "final" release with the current sources before moving to C99.

**Comment by groessler** [comment]
So, yes, sorry, I don't like the move to C99.

I'm often compiling atari800 on "legacy" systems.

I haven't looked at the C99 changes of this PR, but I would help to port them back to C89.

**Comment by MarcinZukowski** [comment]
HI @groessler , thanks. Just curious - what systems exist that still don't support C99? :)
Anyway, I'll take a quick look at the effort to compile it in C89. My most hated thing is all the variables up front... Oh well :)

**Comment by MarcinZukowski** [comment]
OK, so I just committed https://github.com/MarcinZukowski/atari800/commit/7fa87d49b66e97010217c4b75b64c2b7189ebe27 
which makes it compile for me without any c99 switches.

Tested on MacOSX, using clang and gcc-6 (the oldest gcc I could make work).

Note - plenty of warnings still, but only in the "new" code, which is by default disabled.

**Comment by joysfera** [comment]
>My most hated thing is all the variables up front... Oh well :)

please note that the variables don't need to be so much up front (right below function header). You could place them very closely to the place where you need them - just make sure they're the first thing after opening block bracket **{**.

**Comment by MarcinZukowski** [comment]
> please note that the variable declarations don't need to be so much up front (right below function header). You could place them very close to the place where you need them - just make sure they're the first thing after an opening block bracket **{** (at least I hope so).

Yeah, that helps for nested block, but not so much for longish functions.

I'm just trying to understand why we're still stuck no 20+ years old language standard.......

**Comment by joysfera** [comment]
> I'm just trying to understand why we're still stuck no 20+ years old language standard.......

it's like if somebody tried to understand why we bother with emulating of 40+ years old computers.

**Comment by MarcinZukowski** [comment]
Not exactly, but a fair point :)

**Comment by groessler** [comment]
On 6/18/21 11:43 PM, Marcin Zukowski wrote:
> I'm just trying to understand why we're still stuck no 20+ years old 
> language standard....... 


I guess, because "newer" is not necessarily "better". Or so...

**Comment by mikrosk** [comment]
Love your idea! Closing as resolved, feel free to create a PR with required changes. Ideally:
- C89 compatible
- with reasonable amount of commits (no point in squashing everything into one commit)
- every commit compilable
